Basilisk master bc2 bjs2 1599

TrueSamedi opened the pull request.

TrueSamedi committed:
add StageInvitationController
add StageInvitationDto
add StageInvitationMapper
add StageInvitationService
add ValidateStageInvitationException

TrueSamedi committed:
fixed StageInvitationController
fixed StageInvitationMapper
fixed StageInvitationService

TrueSamedi committed:
fixed StageInvitationController
fixed StageInvitationService

TrueSamedi committed:
fixed StageInvitationMapper

TrueSamedi committed:
add StageInvitationControllerTest
add StageInvitationServiceTest

TrueSamedi committed:
fixed StageInvitationControllerTest

TrueSamedi committed:
fixed StageInvitationControllerTest

TrueSamedi committed:
fixed StageInvitationServiceTest

TrueSamedi committed:
fixed StageInvitationServiceTest
fixed StageInvitationService

TrueSamedi committed:
fixed StageInvitationController

PavelAP96 commented on src/main/java/faang/school/projectservice/controller/StageInvitationController.java:
```diff
@@ -0,0 +1,60 @@
+
+    private final StageInvitationService stageInvitationService;
+
+    private void validate(Object field, String message) {
+        if (field == null) throw new ValidateStageInvitationException(message);
```
не очень хороший подход к валидации
сделай через jakarta, примеры есть в сервисах наших
тем более надо проверять не только на null, айдишник, например, не может быть меньше 1

PavelAP96 commented on src/main/java/faang/school/projectservice/controller/StageInvitationController.java:
```diff
@@ -0,0 +1,60 @@
+
+    private final StageInvitationService stageInvitationService;
+
+    private void validate(Object field, String message) {
+        if (field == null) throw new ValidateStageInvitationException(message);
+    }
```
не очень хороший подход к валидации
сделай через jakarta, примеры есть в сервисах наших
тем более надо проверять не только на null, айдишник, например, не может быть меньше 1

PavelAP96 commented on src/main/java/faang/school/projectservice/controller/StageInvitationController.java:
```diff
@@ -0,0 +1,60 @@
+        return stageInvitationService.create(stageInvitationDto);
+    }
+
+    @PostMapping("/accept")
+    public void acceptInvitation(@PathVariable("userId") Long userId,
+                                 @PathVariable("invitationId") Long invitationId) {
```
лучше использовать `@RequestParam`

PavelAP96 commented on src/main/java/faang/school/projectservice/controller/StageInvitationController.java:
```diff
@@ -0,0 +1,60 @@
+        stageInvitationService.accept(userId, invitationId);
+    }
+
+    @PostMapping("/reject")
+    public void rejectInvitation(@PathVariable("userId") Long userId,
+                                 @PathVariable("invitationId") Long invitationId,
+                                 @PathVariable("description") String description) {
```
лучше использовать `@RequestParam `или `@RequestBody` и передавать данные в теле запроса

PavelAP96 commented on src/main/java/faang/school/projectservice/controller/StageInvitationController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @PostMapping("/accept")
+    public void acceptInvitation(@PathVariable("userId") Long userId,
```
в методах контроллера лучше возвращать данные StageInvitationDto
это более информативно и правильно

PavelAP96 commented on src/main/java/faang/school/projectservice/exception/ValidateStageInvitationException.java:
```diff
@@ -0,0 +1,9 @@
+package faang.school.projectservice.exception;
+
+public class ValidateStageInvitationException extends RuntimeException{
+
+    public ValidateStageInvitationException(String message) {
+        super(message);
+    }
```
не обязательно каждый раз создавать свои исключения)
их уже создано огромное количество, поэтому почти всегда можно найти что - то существующее

PavelAP96 commented on src/main/java/faang/school/projectservice/mapper/StageInvitationMapper.java:
```diff
@@ -0,0 +1,20 @@
+
+    StageInvitationDto toDto(StageInvitation stageInvitation);
+
+    List<StageInvitationDto> toDto(List<StageInvitation> stageInvitation);
```
я бы здесь перегрузку метода не использовал и лучше бы назвал toDtoList, например, улучшит читаемость кода

PavelAP96 commented on src/main/java/faang/school/projectservice/service/StageInvitationService.java:
```diff
@@ -0,0 +1,63 @@
+    private final StageInvitationMapper stageInvitationMapper;
+    private final StageInvitationRepository stageInvitationRepository;
+
+    private StageInvitation validate(Long userId, Long invitationId) {
+        StageInvitation invitation = stageInvitationRepository.findById(invitationId);
+
+        if (invitation == null) {
+            throw new ValidateStageInvitationException("Invitation with provided id: " + invitationId + " are not exist");
+        }
+
+        if (invitation.getInvited().getId().equals(userId)) {
+            throw new ValidateStageInvitationException("User " + userId + " not found in invitation");
+        }
+
+        return invitation;
+    }
```
приватные метода класса пишем под основными публичными методами

PavelAP96 commented on src/main/java/faang/school/projectservice/service/StageInvitationService.java:
```diff
@@ -0,0 +1,63 @@
+    private final StageInvitationMapper stageInvitationMapper;
+    private final StageInvitationRepository stageInvitationRepository;
+
+    private StageInvitation validate(Long userId, Long invitationId) {
```
очень странно, что метод называется validate, а возвращает сущность, которую он даже не проверяет
по сигнатуре метода кажется, что он будет валидировать айдишники, а оказывается, что возвращает приглашение
лучше назови getInvitation

PavelAP96 commented on src/main/java/faang/school/projectservice/service/StageInvitationService.java:
```diff
@@ -0,0 +1,63 @@
+
+    public void accept(Long userId, Long invitationId) {
+        StageInvitation invitation = validate(userId, invitationId);
+
```
ты статус поменял, но разве этого достаточно?
если приглашение принято, то надо добавить приглашенного к этапу

PavelAP96 commented on src/main/java/faang/school/projectservice/service/StageInvitationService.java:
```diff
@@ -0,0 +1,63 @@
+        stageInvitationRepository.save(invitation);
+    }
+
+    public List<StageInvitationDto> getAll(Long id) {
+        return stageInvitationMapper.toDto(stageInvitationRepository
+                .findAll()
+                .stream()
+                .filter(stageInvitation -> stageInvitation.getInvited().getUserId() == id)
+                .toList());
+    }
```
в условии задачи сказано с фильтрами
по поводу фильтров есть материалы на платформе

TrueSamedi committed:
delete ValidateStageInvitationException
add StageInvitationFilter
add StageInvitationFilterAuthor
add StageInvitationFilterInvited
add StageInvitationFilterStatus
add TeamMemberService
fixed StageInvitationController
fixed StageInvitationControllerTest
fixed StageInvitationDto
fixed StageInvitationMapper
fixed StageInvitationService
fixed StageInvitationServiceTest

TrueSamedi committed:
fixed StageInvitationFilterAuthor
fixed StageInvitationFilterInvited
fixed StageInvitationFilterStatus
fixed StageInvitationControllerTest
fixed StageInvitationServiceTest

TrueSamedi committed:
fixed StageInvitationDto
fixed StageInvitationController
fixed StageInvitationControllerTest

PavelAP96 commented on src/main/java/faang/school/projectservice/controller/StageInvitationController.java:
```diff
@@ -0,0 +1,47 @@
+    private final StageInvitationService stageInvitationService;
+
+    @PostMapping("/create")
+    public StageInvitationDto createInvitation(@RequestBody @NonNull StageInvitationDto stageInvitationDto) {
```
валидация не очень правильно сделана
у тебя объект может быть не null, но внутри параметры все будут null
посмотри, как сделана валидация в других dto

PavelAP96 commented on src/main/java/faang/school/projectservice/controller/StageInvitationController.java:
```diff
@@ -0,0 +1,47 @@
+
+    private final StageInvitationService stageInvitationService;
+
+    @PostMapping("/create")
```
Использовать глаголы в эндпоинтах нужно только в ситуациях, когда без них не понятно, что за действие будет производиться
в данном случае, используя метод POST, мы уже понимаем, что будет производиться создании приглашения, соответственно нет необходимости использовать /create
можно просто сделать @PostMapping

PavelAP96 commented on src/main/java/faang/school/projectservice/controller/StageInvitationController.java:
```diff
@@ -0,0 +1,47 @@
+        return stageInvitationService.accept(userId, invitationId);
+    }
+
+    @DeleteMapping("/reject")
```
здесь можно без reject
по DELETE итак понятно

PavelAP96 commented on src/main/java/faang/school/projectservice/controller/StageInvitationController.java:
```diff
@@ -0,0 +1,47 @@
+        return stageInvitationService.create(stageInvitationDto);
+    }
+
+    @PostMapping("/accept")
```
можно использовать PutMapping тогда добавлять accept не будет необходимости

PavelAP96 commented on src/main/java/faang/school/projectservice/service/StageInvitationService.java:
```diff
@@ -0,0 +1,84 @@
+                recommendationRequestDtos.add(stageInvitationMapper.toDto(requests));
+            }
+            return recommendationRequestDtos;
+        } else
```
можно не добавлять else и оставить сразу return, потому что у тебя блок if заканчивается return'ом
соответственно даже без else, если код попадает в блок if, то дальше он не пойдет

если оставляешь с else , то добавляй фигурные скобки, без них плохо читается

PavelAP96 commented on src/main/java/faang/school/projectservice/service/StageInvitationService.java:
```diff
@@ -0,0 +1,84 @@
+
+    public List<StageInvitationDto> getAll(Long id, StageInvitationDto filter) {
+        if (filter != null) {
+            Stream<StageInvitation> recommendationRequests = stageInvitationRepository.findAll().stream();
+            for (StageInvitationFilter stageInvitationFilter : stageInvitationFilterList) {
+                if (stageInvitationFilter.isApplicable(filter))
+                    recommendationRequests = stageInvitationFilter.apply(recommendationRequests, filter);
+            }
+            List<StageInvitationDto> recommendationRequestDtos = new ArrayList<>();
+            for (StageInvitation requests : recommendationRequests.toList()) {
+                recommendationRequestDtos.add(stageInvitationMapper.toDto(requests));
+            }
```
а почему бы не написать это все красиво, используя StreamApi

TrueSamedi replied:
но так ведь более читаемее? 
по сути то же самое, просто более подробно расписано

TrueSamedi committed:
add ValidationGroups
fixed StageInvitationDto
fixed StageInvitationController
fixed StageInvitationService

PavelAP96 committed:
Merge branch 'basilisk-master-bc2' into basilisk-master-bc2-BJS2-1599
